AIO - Manage Secrets - adding preconfigure service principal and key vault policies (#118156)
Azure/azure-content: articles/iot-operations/deploy-iot-ops/howto-manage-secrets.md

   | **Redirect URI** | Select **Web** as the platform. You can leave the web address empty. |

1. Select **Register**.

   When your application is created, you are directed to its resource page.

1. Copy the **Application (client) ID** from the app registration overview page. You'll use this value as an argument when running Azure IoT Operations deployment.

Next, give your application permissions for key vault.

1. On the resource page for your app, select **API permissions** from the **Manage** section of the app menu.

1. Select **Add a permission**.

1. On the **Request API permissions** page, scroll down and select **Azure Key Vault**.

1. Select **Delegated permissions**.

1. Check the box to select **user_impersonation** permissions.

1. Select **Add permissions**.

Create a client secret that will be added to your Kubernetes cluster to authenticate to your key vault.

1. On the resource page for your app, select **Certificates & secrets** from the **Manage** section of the app menu.

1. Select **New client secret**.

1. Provide an optional description for the secret, then select **Add**.

> 1. Copy the **Value** and **Secret ID** from your new secret. You'll use these values later below.

Reviewer: ```suggestion
1. Copy the **Value** and **Secret ID** from your new secret. You'll use these values when you run `az iot ops init`.
```

Reviewer: Do you use the Secret ID later? I see just the secret value being used in az iot ops init.

Retrieve the service principal Object Id

> 1. On the **Overview** page for your app, under the section **Essentials**, click on the **Application name** link under **Managed application in local directory**. This opens the Enterprise Application properties. Copy the Object Id to use later.

Reviewer: ```suggestion
1. On the **Overview** page for your app, under the section **Essentials**, click on the **Application name** link under **Managed application in local directory**. This opens the Enterprise Application properties. Copy the Object Id to use when you run `az iot ops init`.
```

### Create an Azure Key Vault

> Create a new Azure Key Vault service and ensure it has the **Permission Model** to Vault access policy.

Reviewer: ```suggestion
Create a new Azure Key Vault service and ensure it has the **Permission Model** set to Vault access policy.
```

```bash
az keyvault create --enable-rbac-authorization false --name "<your unique key vault name>" --resource-group "<the name of the resource group>"
```

> If you have an existing Key Vault, you can change the Permission Model by executing the following:

Reviewer: ```suggestion
If you have an existing key vault, you can change the permission model by executing the following:
```

info: the service is Key Vault, but an individual instance is a key vault

```bash
az keyvault update --name "<your unique key vault name>" --resource-group "<the name of the resource group>" --enable-rbac-authorization false 
```

> You will need the Key Vault Resource ID below, to retrieve it run:

Reviewer: ```suggestion
You will need the Key Vault resource ID when you run `az iot ops init`. To retrieve the resource ID, run:
```

```bash
az keyvault show --name "<your unique key vault name>" --resource-group "<the name of the resource group>" --query id  -o tsv
```

### Set service principal access policy in Azue Key Vault

The newly created service principal needs **Secret** `list` and `get` access policy for the Azure IoT Operations to work with the secret store. 

Run the following to assign **secret** `get` and `list` permissions to the service principal.

```bash

az keyvault set-policy --name "<your unique key vault name>" --resource-group "<the name of the resource group>" --object-id <Object ID copied from Enterprise Application SP in Microsoft Entra ID> --secret-permissions get list --key-permissions get list

> ```

Reviewer: Can remove newlines from within the code block

Author: Updated

> ### Passing in service principal and Azure Key Vault arguments when running Azure IoT Operations deployment

Reviewer: ```suggestion
### Pass service principal and Key Vault arguments to Azure IoT Operations deployment
```

When following the guide [Deploy Azure IoT Operations extensions](./howto-deploy-iot-operations.md?tabs=cli), you will need to pass in additional flags to the `az iot ops init` command in order to use the pre-configured service principal and key vault.

> For example, the following shows how to prepare the cluster for Azure IoT Operations without fully deploying it by using `--no-deploy` flag. You can also run the command without this argument for a default Azure IoT Operations deployment.

Reviewer: ```suggestion
The following example shows how to prepare the cluster for Azure IoT Operations without fully deploying it by using `--no-deploy` flag. You can also run the command without this argument for a default Azure IoT Operations deployment.
```

```bash

> 

Reviewer: Ditto newlines comment

az iot ops init --name "<your unique key vault name>" --resource-group "<the name of the resource group>" \
    --kv-id <Key Vault Resource ID retrieved above> \

>     --sp-app-id <Application registration App ID (client ID) from Microsoft Entra ID above> \

Reviewer: I tried to flag all the steps where you grab values for this final command by saying specifically that they'll be used here. I don't see the step where app ID is retrieved, though.

Author: This is the Application ID from Step 6 in the creation of the Service Principal app registration in Entra ID UI. Maybe we need to point out in that line that this Application (client) ID is visible as such in the Entra ID overview page. Something like 
`--sp-app-id <Application (client) ID from Microsoft Entra ID overview app registration page> `

>     --kv-id <Key Vault Resource ID retrieved above> \
>     --sp-app-id <Application registration App ID (client ID) from Microsoft Entra ID above> \
>     --sp-object-id <Object ID copied from Enterprise Application in Microsoft Entra ID above> \

Reviewer: ```suggestion
    --kv-id <Key Vault Resource ID> \
    --sp-app-id <Application registration App ID (client ID) from Microsoft Entra ID> \
    --sp-object-id <Object ID copied from Enterprise Application in Microsoft Entra ID> \
```

info: in docs, we avoid directional terms like above and below. If you think it's necessary to refer to another section or step, then instead we can use temporal terms like previously, earlier, or later.
